types/wlr_relative_pointer_v1.c

```diff
@@ -0,0 +1,234 @@
+#include <stdlib.h>
+#include <assert.h>
+#include <inttypes.h>
+#include <wlr/util/log.h>
+#include <util/signal.h>
+#include <wlr/types/wlr_relative_pointer_v1.h>
+#include <wlr/types/wlr_seat.h>
+#include "wayland-util.h"
+#include "wayland-server.h"
+#include "relative-pointer-unstable-v1-protocol.h"
+
+#define RELATIVE_POINTER_MANAGER_VERSION 1
+
+static const struct zwp_relative_pointer_manager_v1_interface relative_pointer_manager_v1_impl;
+static const struct zwp_relative_pointer_v1_interface relative_pointer_v1_impl;
+
+
+/**
+ * helper functions
+ */
+
+struct wlr_relative_pointer_v1 *relative_pointer_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_relative_pointer_v1_interface,
+		&relative_pointer_v1_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+
+static struct wlr_relative_pointer_manager_v1 *relative_pointer_manager_from_resource(struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource, &zwp_relative_pointer_manager_v1_interface,
+		&relative_pointer_manager_v1_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+
+/**
+ * relative_pointer handler functions
+ */
+
+static void relative_pointer_v1_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_relative_pointer_v1 *relative_pointer =
+		relative_pointer_from_resource(resource);
+	wlr_signal_emit_safe(&relative_pointer->events.destroy, relative_pointer);
+
+	wl_list_remove(wl_resource_get_link(resource));
+	free(relative_pointer);
+}
+
+
+static void relative_pointer_v1_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	struct wlr_relative_pointer_v1 *relative_pointer =
+		relative_pointer_from_resource(resource);
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 %p released by client %p",
+		relative_pointer, client);
+
+	wl_resource_destroy(resource);
+}
+
+
+/**
+ * relative_pointer_manager handler functions
+ */
+
+static void relative_pointer_manager_v1_handle_resource_destroy(struct wl_resource *resource) {
+	wl_list_remove(wl_resource_get_link(resource));
+}
+
+
+static void relative_pointer_manager_v1_handle_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+
+	wlr_log(WLR_DEBUG, "relative_pointer_v1 manager unbound from client %p",
+		client);
+}
+
+
+static void relative_pointer_manager_v1_handle_get_relative_pointer(struct wl_client *client,
+		struct wl_resource *resource, uint32_t id, struct wl_resource *pointer) {
+	struct wlr_seat_client *seat_client =
```

Comment:
We should listen for the seat being destroyed, and make the resource inert (see CONTRIBUTING.md) when this happens. Otherwise we'll have a dangling invalid pointer to a `wlr_seat`.

Reply:
Should we also listen for the seat_client being destroyed as well?

Reply:
I don't think so, because it's fine to create a `wl_seat`, create a `wl_pointer`, create a relative pointer, and then destroy the seat and continue to use both pointers.